### PR TITLE
[megatron, vllm, recipe] feat: add Qwen2.5-0.5B CISPO training on Ascend

### PR DESCRIPTION
**Pull request by RordChang**
### What does this PR do?

Adds a Qwen2.5-0.5B-Instruct CISPO training example using Megatron for actor/reference training and vLLM-Ascend for rollout on Ascend NPUs.

- Adds a canonical Megatron + vLLM-Ascend launcher with environment-variable overrides, automatic checkpoint resume, and timestamped logs.
- Connects the registered CISPO policy loss to GRPO advantages with lower/upper clipping of `10.0 / 0.2` and token-mean aggregation.
- Adds a concise entry point to the existing CISPO README and links to the full environment, logs, and 100-step results.

Related:
https://github.com/verl-project/verl-ascend-recipe/issues/17

### Checklist Before Starting

- [x] Searched relevant open PRs:
  - [CISPO Ascend](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+CISPO+Ascend)
  - [CISPO Megatron](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+CISPO+Megatron)
  - PR #4508 already provides the registered CISPO policy loss; no open PR covers this Qwen2.5-0.5B, Megatron, and vLLM-Ascend training combination.
- [x] Formatted the title as `[{modules}] {type}: {description}`.

### Test

- `pre-commit run --all-files --show-diff-on-failure --color=always`: passed.
- CISPO objective and registry CPU validation: 2 passed.
- `bash -n examples/cispo_trainer/run_qwen2_5_0_5b_megatron.sh`: passed.

### Experiment Results

Platform: 4 x Ascend 910B.

| Metric | Result |
| --- | ---: |
| Completed training steps | 100/100 |
| Reward mean, first 10 steps | 0.010156 |
| Reward mean, last 10 steps | 0.281250 |
| Reward change | +0.271094 |
| Reward linear slope | +0.00388862/step |
| Mean throughput per NPU | 258.8282 token/s/NPU |
| Mean total throughput | 1035.3128 token/s |

Full redacted 100-step training log:
[training_100step_sanitized.log](https://github.com/user-attachments/files/31128595/training_100step_sanitized.log)

### API and Usage Example

```bash
python3 examples/data_preprocess/gsm8k.py \
    --local_save_dir /path/to/data/gsm8k

MODEL_PATH=/path/to/Qwen2.5-0.5B-Instruct \
DATA_ROOT=/path/to/data \
NPUS_PER_NODE=4 \
TOTAL_TRAINING_STEPS=100 \
bash examples/cispo_trainer/run_qwen2_5_0_5b_megatron.sh
```

Additional Hydra overrides can be appended to the command.

### Design & Code Changes

- The existing `cispo` policy loss remains the single implementation of the objective.
- The Megatron actor calls the policy loss through the common PPO loss path; actor/reference training uses TP=2 and BF16.
- vLLM-Ascend rollout uses TP=2, four responses per prompt, chunked prefill, prefix caching, and bucketed weight synchronization.
- The launcher exposes model, data, parallelism, batch size, checkpoint, and logging settings through environment variables.
- The existing CISPO README only gains the new entry point and a link to complete results.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] Applied pre-commit checks.
- [x] Updated documentation.
- [x] CPU validation results are reported above.
- [ ] Request CI after the PR is ready.
- [x] Recipe submodule update: N/A; this PR does not modify the submodule.

AI assistance: OpenAI Codex was used for codebase inspection and preparation of the implementation, validation, and documentation. The submitter reviewed the changes and validation results.

### CODE REVIEW

**Comment by fangweii1** [comment]
> ### What does this PR do?
> Adds a Qwen2.5-0.5B-Instruct CISPO training example using Megatron for actor/reference training and vLLM-Ascend for rollout on Ascend NPUs.
> 
> * Adds a canonical Megatron + vLLM-Ascend launcher with environment-variable overrides, automatic checkpoint resume, and timestamped logs.
> * Connects the registered CISPO policy loss to GRPO advantages with lower/upper clipping of `10.0 / 0.2` and token-mean aggregation.
> * Adds CPU coverage for clipped-ratio behavior, stop-gradient semantics, response masking, and policy-loss registration.
> * Adds a concise entry point to the existing CISPO README and links to the full environment, logs, and 100-step results.
> 
> Related: [verl-project/verl-ascend-recipe#17](https://github.com/verl-project/verl-ascend-recipe/issues/17)
> 
> ### Checklist Before Starting
> * [x]  Searched relevant open PRs:
>   
>   * [CISPO Ascend](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+CISPO+Ascend)
>   * [CISPO Megatron](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+CISPO+Megatron)
>   * PR [[algo] feat: add cispo #4508](https://github.com/verl-project/verl/pull/4508) already provides the registered CISPO policy loss; no open PR covers this Qwen2.5-0.5B, Megatron, and vLLM-Ascend training combination.
> * [x]  Formatted the title as `[{modules}] {type}: {description}`.
> 
> ### Test
> * `pre-commit run --all-files --show-diff-on-failure --color=always`: passed
> * CISPO objective and registry CPU tests: 2 passed
> * `bash -n examples/cispo_trainer/run_qwen2_5_0_5b_megatron.sh`: passed
> * 100-step training:
>   
>   * reward first/last 10-step mean: 0.010156 / 0.281250
>   * reward linear slope: 0.00388862/step
>   * mean total throughput on 4 × Ascend 910B4: 1035.3128 token/s
> 
> Training artifacts: [verl-project/verl-ascend-recipe#17 (comment)](https://github.com/verl-project/verl-ascend-recipe/issues/17#issuecomment-5308881002)
> 
> ### API and Usage Example
> ```shell
> python3 examples/data_preprocess/gsm8k.py \
>     --local_save_dir /path/to/data/gsm8k
> 
> MODEL_PATH=/path/to/Qwen2.5-0.5B-Instruct \
> DATA_ROOT=/path/to/data \
> NPUS_PER_NODE=4 \
> TOTAL_TRAINING_STEPS=100 \
> bash examples/cispo_trainer/run_qwen2_5_0_5b_megatron.sh
> ```
> 
> Additional Hydra overrides can be appended to the command.
> 
> ### Design & Code Changes
> * The existing `cispo` policy loss remains the single implementation of the objective.
> * The Megatron actor calls the policy loss through the common PPO loss path; actor/reference training uses TP=2 and BF16.
> * vLLM-Ascend rollout uses TP=2, four responses per prompt, chunked prefill, prefix caching, and bucketed weight synchronization.
> * The launcher exposes model, data, parallelism, batch size, checkpoint, and logging settings through environment variables.
> * The existing CISPO README only gains the new entry point and a link to complete results.
> 
> ### Checklist Before Submitting
> * [x]  Read the Contribute Guide.
> * [x]  Applied pre-commit checks.
> * [x]  Updated documentation.
> * [x]  Added CPU tests to the existing policy-loss test suite.
> * [ ]  Request CI after the PR is ready.
> * [x]  Recipe submodule update: N/A; this PR does not modify the submodule.
> 
> AI assistance: OpenAI Codex was used for codebase inspection and preparation of the implementation, tests, and documentation. The submitter reviewed the changes and validation results.

请完善一下 1. 不需要提交test相关的代码  2. 提交实验数据

**Comment by RordChang** [comment]
@fangweii1 已按反馈完成修改：test 相关代码已删除，实验数据和完整脱敏训练日志链接已补充至 PR 正文。

**Comment by wucong25** [comment]
统一先提到verl-ascend-recipe：[verl-ascend-recipe](https://github.com/verl-project/verl-ascend-recipe/tree/main/verl_ascend_practice)哈，此pr暂时关闭
